packages/js/product-editor/src/components/prepublish-panel/visibility-section/visibility-section.tsx

```diff
@@ -0,0 +1,186 @@
+import { __ } from '@wordpress/i18n';
+import { Product, ProductCatalogVisibility } from '@woocommerce/data';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEntityProp } from '@wordpress/core-data';
+import { useInstanceId } from '@wordpress/compose';
+import { recordEvent } from '@woocommerce/tracks';
+import {
+	BaseControl,
+	CheckboxControl,
+	// @ts-expect-error `__experimentalInputControl` does exist.
+	__experimentalInputControl as InputControl,
+	PanelBody,
+} from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { VisibilitySectionProps } from './types';
+import { store as productEditorUiStore } from '../../../store/product-editor-ui';
+import { TRACKS_SOURCE } from '../../../constants';
+
+export function VisibilitySection( { productType }: VisibilitySectionProps ) {
+	const [ catalogVisibility, setCatalogVisibility ] = useEntityProp<
+		Product[ 'catalog_visibility' ]
+	>( 'postType', productType, 'catalog_visibility' );
+
+	const [ reviewsAllowed, setReviewsAllowed ] = useEntityProp<
+		Product[ 'reviews_allowed' ]
+	>( 'postType', productType, 'reviews_allowed' );
+
+	const [ postPassword, setPostPassword ] = useEntityProp< string >(
+		'postType',
+		productType,
+		'post_password'
+	);
+
+	const { requirePassword } = useDispatch( productEditorUiStore );
+
+	const isPasswordRequired: boolean = useSelect( ( select ) => {
+		return select( productEditorUiStore ).isPasswordRequired();
+	}, [] );
+
+	const postPasswordId = useInstanceId(
+		BaseControl,
+		'post_password'
+	) as string;
+
+	function handleVisibilityChange(
+		selected: boolean,
+		visibility: ProductCatalogVisibility
+	) {
+		if ( selected ) {
+			if ( catalogVisibility === 'visible' ) {
+				setCatalogVisibility( visibility );
+				recordEvent( 'product_prepublish_panel', {
+					source: TRACKS_SOURCE,
+					action: 'catalog_visibility',
+					visibility: catalogVisibility,
+				} );
+				return;
+			}
+			setCatalogVisibility( 'hidden' );
+		} else {
+			if ( catalogVisibility === 'hidden' ) {
+				if ( visibility === 'catalog' ) {
+					setCatalogVisibility( 'search' );
+					recordEvent( 'product_prepublish_panel', {
+						source: TRACKS_SOURCE,
+						action: 'catalog_visibility',
+						visibility: catalogVisibility,
+					} );
+					return;
+				}
+				if ( visibility === 'search' ) {
+					setCatalogVisibility( 'catalog' );
+					recordEvent( 'product_prepublish_panel', {
+						source: TRACKS_SOURCE,
+						action: 'catalog_visibility',
+						visibility: catalogVisibility,
+					} );
+					return;
+				}
+				return;
+			}
+			setCatalogVisibility( 'visible' );
+			recordEvent( 'product_prepublish_panel', {
+				source: TRACKS_SOURCE,
+				action: 'catalog_visibility',
+				visibility: catalogVisibility,
+			} );
+		}
+	}
+
+	return (
+		<PanelBody
+			initialOpen={ false }
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore We need to send an Element.
```

Comment:
Considering how [the `title` prop is being used](https://github.com/WordPress/gutenberg/blob/62a7316473b74911670fc28fcfe3b927a7fbb808/packages/editor/src/components/post-publish-panel/prepublish.js#L112-L120) on the GB pre-publish panel, I think that this [TS rule](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/panel/types.ts#L93) will change. Otherwise we should modify it in the future.